### PR TITLE
allows to add a new tagname hitting enter

### PR DESCRIPTION
**Pull request by tobibeer**
demo: http://2592.tiddlyspot.com
discussion: https://groups.google.com/forum/#!topic/tiddlywiki/wqQ8jPYG-X4
requires: #2593 for proper layout (using a span wrapper rather than a div)

### CODE REVIEW

File: themes/tiddlywiki/vanilla/base.tid [hunk 1]
```diff
@@ -2186,3 +2186,7 @@ body.tc-dirty span.tc-dirty-indicator, body.tc-dirty span.tc-dirty-indicator svg
 	background: #f00;
 	color: #fff;
 }
+
+.tc-keyboard {
+	display: inline;
```
Comment (pmario) [inline themes/tiddlywiki/vanilla/base.tid]: why do we need this?

Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: because the keyboard widget, for some reason, creates a **div** wrapper

Comment (Jermolene) [inline themes/tiddlywiki/vanilla/base.tid]: > for some reason

It's pretty obvious: we need to DIV so that we've got somewhere to attach the keyboard event handlers.

Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: > for some reason

was referring to it having to be a block-level element... but other than that, I figured you didn't want to attach the keyboard handling to some parent element

Comment (Jermolene) [inline themes/tiddlywiki/vanilla/base.tid]: > was referring to it having to be a block-level element

Ah, I see. Other widgets (eg button and reveal) have established the convention of a `tag` attribute that allows the default generated HTML element to be overridden.

Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: @Jermolene, would you be fine with me to add the necessary changes to this PR?

Comment (Jermolene) [inline themes/tiddlywiki/vanilla/base.tid]: > would you be fine with me adding the necessary changes to this PR

Better to make it a separate PR

Comment (tobibeer) [inline themes/tiddlywiki/vanilla/base.tid]: > Better to make it a separate PR

@Jermolene, done, see: #2593

will revert changes to vanilla/base

**Comment by pmario** [comment]
LGTM

**Comment by danielo515** [comment]
Tested on android 6 chrome and it works. 
Regards

**Comment by tobibeer** [comment]
@Jermolene, unfortunately, there's no comment as to what needs work.

Apart from the dependency to #2592, I'm not sure there was anything to do here? I don't know of a spot in the documentation that would need some mentioning.

Not sure, perhaps you were considering adding it to the ControlPanel shortcuts, but I really don't see how this fits there.

**Comment by pmario** [comment]
link to the docs: http://tiddlywiki.com/#Creating%20and%20editing%20tiddlers

eg: replace `Then click the add button to add the tag to the tiddler.`
with:  `Then ''click the add'' button or hit the <kbd>Enter</kbd> key, to add the tag to the tiddler. `

**Comment by pmario** [comment]
>Apart from the dependency to #2592

This is 2592??

**Comment by tobibeer** [comment]
> link to the docs: http://tiddlywiki.com/#Creating%20and%20editing%20tiddlers

Thanks for the pointers, @pmario, adding your proposed changes...

**Comment by Jermolene** [comment]
I think my original comments have been lost, but there's still the problem with removing the conditional in `SimpleEngine.prototype.setText()`.

**Comment by tobibeer** [comment]
@Jermolene, can you be specific about what that problem is? All I remember is that it needed to be changed or otherwise fail.

**Comment by Jermolene** [comment]
Hi @tobibeer the current code for `SimpleEngine.prototype.setText()` avoids updating the text of the text box if it contains the focus. You can easily see why it is needed if you take the current repo, and change setText to remove the conditional test. You'll see that typing in the search field works as expected until you try to move the cursor to the start of the field and type a character; each character results in the text box being updated which moves the cursor straight to the end of the text box.

```
SimpleEngine.prototype.setText = function(text,type) {
	if(!this.domNode.isTiddlyWikiFakeDom) {
		this.domNode.value = text;
		// Fix the height if needed
		this.fixHeight();
	}
};
```

**Comment by tobibeer** [comment]
@Jermolene, thanks for clarifying

How about extending setText so that certain keystrokes like enter or tab force a refresh? This should be fine since we're only in a basic text input. Shouldn't it?

**Comment by tobibeer** [comment]
When in a simple text input, [SimpleEngine.prototype.handleInputEvent](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/editor/engines/simple.js#L112-L116) should probably not be invoking [EditTextWidget.prototype.saveChanges](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/editor/factory.js#L243-L248) but rather stop doing anything when we hit ENTER or TAB.

**Comment by tobibeer** [comment]
Mhhh, for some reason, the event passed to [SimpleEngine.prototype.handleInputEvent](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/editor/engines/simple.js#L112-L116) doesn't come with a `keyCode` I could check in order to decide whether to update / refresh the widget or not?

**Comment by tobibeer** [comment]
In other words, this won't work:

```
SimpleEngine.prototype.handleInputEvent = function(event) {
	if(
		this.widget.editTag === "textarea" ||
	  	this.widget.editTag !== "textarea" && [9,13].indexOf(event.keyCode) < 0
	) {
		this.widget.saveChanges(this.getText());
		this.fixHeight();
	}
	return true;
};
```

**Comment by tobibeer** [comment]
The problem appears to be that we are using the oninput / input event for the input element which doesn't support `keyCode`.

**Comment by tobibeer** [comment]
I reverted setText back, so now the input field will not be cleared when the tag is added.

**Comment by tobibeer** [comment]
I reverted setText back, so now the input field will not be cleared when the tag is added.

**Comment by Jermolene** [comment]
Hi @tobibeer please can you resolve the merge conflict?

**Comment by tobibeer** [comment]
will need to investigate more what the tag-picker macro is

**Comment by tobibeer** [comment]
as said, the behavior now is not perfect, since the input is not cleared

the reason is that this:

`<$action-deletetiddler $tiddler="$:/temp/NewTagName"/>`

does not get executed after the inner edittext refreshed and so the inner edittext ends up setting the same text value back to the input

demo is updated accordingly

**Comment by pmario** [comment]
>as said, the behavior now is not perfect, since the input is not cleared

but I still think, it's a much requested improvement. You can hit [enter] [ctrl] [a] and go on typing. ... Not perfect, but no mouse needed :)

**Comment by tobibeer** [comment]
> but I still think, it's a much requested improvement. You can hit [enter] [ctrl] [a] and go on typing. ... Not perfect, but no mouse needed :)

You're right, compared to before, it is an improvement I would probably use all the time, perfect or not

**Comment by duarteframos** [comment]
This is pretty cool, awesome usability improvement!
Could similar code be recycled to make the _Add Link_ or _Add Image_ Editor Toolbar buttons also work with the [Enter] key??

**Comment by pmario** [comment]
@duarteframos IMO you should open a new issue that links to this PR, or new PR. Your post here may be forgotten very fast.

**Comment by duarteframos** [comment]
Will do, thanks for the suggestion.
Reported as https://github.com/Jermolene/TiddlyWiki5/issues/2668
